Final version for Kontrol pausability proofs

JuanCoRo opened the pull request:
<!-- Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md -->

**Description**

Building on #9183, this PR wraps up the existing pausability proofs by adding the following:

- Native symbolic support for arguments of type `bytes` and `bytes[]`
- Replace `mockCall` workaround with the `vm.mockCall` cheatcode
- Replace `startPrank` workaround with the appropriate `vm.prank` cheatcode

**Tests**

No additional tests are added so far. This PR modifies the existing proofs to include all the latest and necessary improvements.

**Additional context**

Before this PR, a number of workarounds were in place to have these proofs pass. These workarounds didn't threaten the soundness of the proofs, but they affected long-term maintainability, execution performance, and the generality of the proofs.

JuanCoRo committed:
Update tests to native symbolic `bytes` and `bytes[]`

Commit:
adding lemmas

JuanCoRo committed:
`run-kontrol.sh`: add more sensible parameters

JuanCoRo committed:
Change `startPrank` by `prank`

JuanCoRo committed:
Replace `mockCall` workaround with `vm.mockCall`

JuanCoRo committed:
Make `bytes` length symbolic

JuanCoRo committed:
`KontrolUtils`: remove symbolic workarounds

JuanCoRo committed:
`run-kontrol.sh`: add `prove_proveWithdrawalTransaction_paused`

JuanCoRo committed:
`forge fmt`

JuanCoRo committed:
`versions.json`: bump Kontrol from `0.1.127` to `0.1.156`

JuanCoRo committed:
Remove `ASSUME` comments

mds1 commented on packages/contracts-bedrock/test/kontrol/pausability-lemmas.k:
```diff
@@ -1,12 +1,910 @@
+ // --------------------------------------
+    rule <k> runLemma(T) => doneLemma(T) ... </k>
+
+    // We need to enforce some limit on the length of bytearrays
+    // and indices into bytearrays in order to avoid chop-reasoning
+    syntax Int ::= "maxBytesLength" [alias]
+    rule maxBytesLength => 9223372036854775808
```
Why do we want to avoid chop-reasoning and why is 2^63 the max length? In general, looks like a lot of lemmas were added and I'd love to understand why they are all needed, towards the goal of being able to maintain the proofs ourselves. Some of the lemmas, like commutativity, feel like they're generic enough that I'm surprised it needs to be specified here

JuanCoRo replied:
Chop-reasoning means reasoning about `mod 2^256`. `chop` is defined as `modInt pow265` [here (at the bottom)](https://github.com/runtimeverification/evm-semantics/blob/master/kevm-pyk/src/kevm_pyk/kproj/evm-semantics/word.md#range-of-types). @PetarMax can confirm, but having 2⁶³ be the max length is required by the compiler-introduced checks on `bytes` parameters. Chop (modulo) reasoning is notoriously difficult for automated tools in general, so if we have `x <= 2^63` we can transform `x mod 2^256` into `x`, which is more easy to reason about.
As per the lemmas, some of the work left to do is to see how many can we remove and still have the proof pass.
Regarding the documentation of these, I'll change `pausability-lemmas.k` into `pausability-lemmas.md` to be able to thoroughly document the lemmas in the same style that you can see the above linked file

mds1 commented on packages/contracts-bedrock/test/kontrol/scripts/run-kontrol.sh:
```diff
@@ -105,8 +105,8 @@ regen=
 #################################
-# Missing: OptimismPortalKontrol.prove_proveWithdrawalTransaction_paused
-test_list=( "OptimismPortalKontrol.prove_finalizeWithdrawalTransaction_paused" \
+test_list=( "OptimismPortalKontrol.prove_proveWithdrawalTransaction_paused" \
+            "OptimismPortalKontrol.prove_finalizeWithdrawalTransaction_paused" \
```
Since we now run all tests by default, can we remove the need for the `--match-test` flag all together, and update this so any extra args passed to the script are forwarded to `kontrol prove`? This wasy I can do `./run-kontrol.sh --match-test prove_proveWithdrawalTransaction_paused` to run a single test

JuanCoRo replied:
Sure thing! I would be more inclined to keep it as it is, but if you prefer feeding the tests to `run-kontrol.sh` directly it certainly can be done! Also, do you want to add `--match-test` for each test, or just the name of the tests?
Also also, should the behavior when no parameters are passed (`./run-kontrol.sh dev`) be to execute all the tests, or throw an error? For CI purposes it would seem best to do the former

mds1 replied:
My thinking was that the best UX for `./run-kontrol.sh` is to match the UX of `forge test` as much as possible, so by default we'd run all detected tests, and pass flags to filter. So the assumption I made in my comment above is that by default `kontrol prove` would find all `prove_*` tests and run the, but maybe that's incorrect?

JuanCoRo replied:
That seems reasonable to me! But I would have the following 3 behaviors instead of 2:
- `./run-kontrol.sh [container/local/dev] name_of_test`: execute `name_of_test` in the mode specified
- `./run-kontrol.sh [container/local/dev]`: execute all tests specified in `run.kontrol.sh`
- `./run-kontrol.sh [container/local/dev] all`: execute `kontrol prove`, matching all `prove_*`, `test_*` and `check_*` functions

Does this seem alright? The main point is to have some control (no pun intended lol) at the script level of what proofs are being executed and not rely entirely on passing arguments to the script. This can end up in a particularly long command if we're feeding it around 15 proofs

Also, it turns out that for the moment, `kontrol prove` without any `--match-test` will only match `test_*` functions, but we're already addressing this, so it will land soon in Kontrol

mds1 replied:
> The main point is to have some control (no pun intended lol) at the script level of what proofs are being executed and not rely entirely on passing arguments to the script. This can end up in a particularly long command if we're feeding it around 15 proofs

Ah, this seems to lead to the difference in our preferences—why would we feed 15 proofs as inputs to that script? Presumably that's because we e.g. have 16 proofs and want to skip one for now for some reason. Going back to matching `forge test` UX, if there are certain proofs we explicitly don't want to run, then you'd `vm.skip(true)` in the test body. Then the resulting UX would be:
- `./run-kontrol.sh [container/local/dev]`: execute `kontrol prove`, matching all `prove_*`, `test_*` and `check_*` functions
- `./run-kontrol.sh [container/local/dev] [--match-test $test] [--match-path $path] [--match-contract $contract]`: match tests based on the provided flags

Your proposal is also ok though if there's a reason mine doesn't work

JuanCoRo replied:
The only reason I'd like to have some control at the script level still is when developing proofs because we might want to exclude/include some, which may lead to an arbitrary subset of the proofs being executed. I think it would be a bit cumbersome to edit a large command when doing proof development or testing out stuff rather than just commenting some lines in `run-kontrol.sh`.
I don't see anything wrong with what you're suggesting! So, maybe we could have the two cases you're describing plus
- `./run-kontrol.sh [container/local/dev] script`: execute all tests specified in `run-kontrol.sh` (it doesn't need to be specifically `script`)

What do you think? If you would rather not have this option I'm describing, I'm okay with leaving it out

mds1 replied:
I know you mentioned you're working on some docs around best practices for developing proofs so I may not be up to speed here. My thought was you'd typically just run one proof, one file, or one contract at a time—all of which just need a single CLI arg.

Ultimately I don't feel too strongly as long as it's well documented. Matching forge UX felt like the right direction to me, but if proof writing is different enough than standard test writing, perhaps that is not actually optimal. I do think the `script` arg may be a good compromise though!

JuanCoRo replied:
Ultimately, the `run-kontrol.sh` script must fit you, so if you find it more intuitive/consistent to work with `forge` UX, that's fine with me! I do think it's a good idea! It's just that, from what I'm used to do in proof development, I see keeping the `script` option useful.
A case for having the script option is, for instance, to execute all proofs that only have a `bytes` argument and no `bytes[]`. In that case, we're only excluding one proof, which we can also do by adding `vm.skip()` to it, but I think I'd lean more toward not modifying the code if we could just modify the execution parameters

mds1 replied:
Makes sense, let's keep the `script` option then. So I think the final state we have is:
- no args: run all tests
- script arg: run the ones specified in the script
- otherwise, forward all matching args like `--match-test`

JuanCoRo replied:
Agreed!

mds1 commented on packages/contracts-bedrock/test/kontrol/scripts/run-kontrol.sh:
```diff
@@ -120,8 +120,8 @@ done
 #########################
-max_depth=1000000
-max_iterations=1000000
+max_depth=10000
+max_iterations=10000
```
Why were these changed and how do you know what to set them at in the first place? I know you said you were working on some documentation for all these options, so perhaps that will answer this

JuanCoRo replied:
These were changed because the previous value was set to not produce any nodes while executing the summary contract with Kontrol. 10000 is sort of the standard that we use. When observing the KCFGs, usually the steps between the nodes never exceed 10k (~8k is the max I've observed for our proofs). This means that if we were to set a lower value (e.g., 1k) Kontrol would create nodes after that many steps of execution, but this would increase execution time. Here, the rationale is that there are not more than 10k steps between one node and the following in normal conditions. We set it to 10k as a way of measuring and inspecting when this happens. [Here's the documentation](https://docs.runtimeverification.com/kontrol/cheatsheets/kontrol-cheatsheet#flags) we have on these flags so far, which I feel might be improved. Any feedback welcome!

mds1 replied:
I see, thanks for explaining! I read the linked docs—can we add `--counterexample-information` as a flag? I recall that UX was very helpful for failures, but unsure if there's any downsides to using that as the default

JuanCoRo replied:
Actually, your comment suggests that we should improve our docs. Flags that exist with a `no` version (e.g., `--counterexample-information` and `--no-counterexample-information`) always have the default behavior as the flag without the `no`. So, in this case, unless otherwise specified, Kontrol runs by default with the `--counterexample-information` enabled

EDIT: Issue tracking this https://github.com/runtimeverification/kontrol/issues/379

mds1 replied:
Oh perfect! Though I don't see a `--no-counterexample-information` in the linked docs 

Either way will resolve this thread

mds1 commented on packages/contracts-bedrock/test/kontrol/proofs/utils/KontrolUtils.sol:
```diff
@@ -4,78 +4,7 @@ pragma solidity 0.8.15;
-}
-
 /// @notice Tests inheriting this contract cannot be run with forge
 abstract contract KontrolUtils is KontrolCheats {
```
Feels like this file can be deleted now to simplify things? Regular tests would then directly inherit from `KontrolCheats`, and they'll already have `vm`

JuanCoRo replied:
As of now, the whole point of `KontrolUtils` is just to provide `vm` and `KontrolCheats`, since in `KontrolCheats` we don't have a variable named `vm` (we have `kevm`) to avoid possible conflicts.
We could add to each test `import { Vm } from "forge-std/Vm.sol";` and `Vm internal constant vm = Vm(address(uint160(uint256(keccak256("hevm cheat code")))));`, but perhaps this way is cleaner?

mds1 replied:
Oh I see, the `.k.sol` tests do not inherit from the regular foundry `Test` contract which includes `vm`. If it doesn't break anything, inheriting from that to get `vm` would be ideal, e.g. we'd have `contract L1ERC721BridgeKontrol is Test, DeploymentSummary, KontrolCheats`

JuanCoRo replied:
I'll test adding `Test` as you suggest! The rationale behind excluding it was to declutter the `.k.sol` files as much as possible, particularly not including the `Test` contract

mds1 replied:
Could we add a comment explaining why we want to keep this file? Quoting from slack:

> As per having `Test` as a dependency of the tests and doing away with `KontrolUtils`. After discussing it internally, we think it'd be better to to keep `KontrolUtils` in favor of `Test`. The reason being that the summary for the `copy-memory-to-memory` function is dependent on the bytecode of the test contract. It has some flexibility, but adding `Test` as a dependency would decrease the robustness of the tests, since big changes to `Test` could impact said summary

mds1 replied:
Bumping this comment

JuanCoRo replied:
Done!

mds1 commented on packages/contracts-bedrock/test/kontrol/proofs/L1CrossDomainMessenger.k.sol:
```diff
@@ -19,25 +19,18 @@ contract L1CrossDomainMessengerKontrol is DeploymentSummary, KontrolUtils {
         uint256 _value,
-        uint256 _gas
+        uint256 _gas,
+        bytes calldata _message
```
So this has a max length of 2^63 because of `maxBytesLength` in the lemmas file? Does that affect performance at all? Realistically we're bound by gas costs here—if everything else was free, we'd have 30M gas to spend on all zero bytes of calldata at 4 gas per byte, giving us 30M/4 = 7.5M bytes in length, meaning we could safely shrink that max length

JuanCoRo replied:
AFAIK the max size of the `bytes` parameter doesn't affect the performance. I'll shrink that max length and see if we win some performance. What causes a bit of a slow down is having a symbolic length (capped at 2⁶³) vs a concrete length, although by what I've observed it's just a couple of minutes on my machine

mds1 replied:
Following up here too to see if we should shrink the max length

JuanCoRo replied:
After discussing internally, shrinking the max length of the `bytes` parameters would not yield any performance benefits, so there is no harm in having the most general (although, realistically, overkill) bound. Note that the `maxBytesLength` is used only for the `copy_memory_to_memory` summary, which is relevant for `bytes[]` parameters. The `2^63` length bound is only applied to symbolic `bytes` parameters that are not elements of a `bytes[]`

mds1 commented on packages/contracts-bedrock/test/kontrol/proofs/OptimismPortal.k.sol:
```diff
@@ -20,78 +20,35 @@ contract OptimismPortalKontrol is DeploymentSummary, KontrolUtils {
-    /// TODO: Replace struct parameters and workarounds with the appropriate
-    /// types once Kontrol supports symbolic `bytes` and `bytes[]`
-    /// Tracking issue: https://github.com/runtimeverification/kontrol/issues/272
+    /// @custom:kontrol-length-equals _withdrawalProof: 10,
+    /// @custom:kontrol-length-equals _withdrawalProof[]: 600,
+    /// @custom:kontrol-length-equals data: 1000,
```
Ah neat, looks like these comment directives are parsed to provide bounds on the inputs? Though I'm not sure I fully understand the syntax. `_withdrawalProof: X` specifies the exact array length and `_withdrawalProof[]: Y` specifies the max element length? What about `data: Z`? 

If so, maybe something like this is more clear?
```
/// @custom:kontrol-array-length-equals _withdrawalProof: 10,
/// @custom:kontrol-element-length-equals _withdrawalProof: 600,
```

Either way, per your slack comment, it definitely would be ideal if we can instead change those directives to be min/max *bounds* on the parameters instead of exact amounts, this way kontrol can run all required paths without us having to duplicate the test code for each condition

JuanCoRo replied:
> `_withdrawalProof: X` specifies the exact array length and `_withdrawalProof[]: Y` specifies the max element length? What about `data: Z`?

Yes, that is correct for `X`! For `Y`, it also specifies the exact length. For `Z`, `data` is the name of the `bytes` field in the struct `Types.WithdrawalTransaction`. Currently, the way of specifying concrete lengths for parameters in structs is by addressing the parameter's name (although this might change in the future).

Regarding the bounds on the parameters, after looking into it, it is not currently possible to have bounds on the length of the elements of `bytes[]`, only concrete lengths. Adding bounds on these, as well as adding bounds for `bytes[]` parameter's length, is quite convoluted and is currently a research item with no concrete ETA on completion. Eventually, we should be able to get there, but given the complexity of the task, I'd say it's something we cannot count on for the moment. I hope that's okay!

As per the renaming of the natspec keys, we're internally discussing it, but it makes a lot of sense to have it the way you're suggesting! Will update you as soon as I have more news on this end

mds1 replied:
Got it, that all sounds good, thanks!

JuanCoRo replied:
We did manage to get the natspec refactoring on time for this PR! [Here's the commit](https://github.com/ethereum-optimism/optimism/pull/9530/commits/40abc60207b466f797a52a7c074854d1caecd15f) that updates the needed comments. After internally discussing, this is the formulation to which we arrived. Let me know if you've got any comments or concerns!

mds1 replied:
That LGTM, thank you!

mds1 commented on packages/contracts-bedrock/test/kontrol/proofs/OptimismPortal.k.sol:
```diff
@@ -20,78 +20,35 @@ contract OptimismPortalKontrol is DeploymentSummary, KontrolUtils {
-    /// TODO: Replace struct parameters and workarounds with the appropriate
-    /// types once Kontrol supports symbolic `bytes` and `bytes[]`
-    /// Tracking issue: https://github.com/runtimeverification/kontrol/issues/272
+    /// @custom:kontrol-length-equals _withdrawalProof: 10,
+    /// @custom:kontrol-length-equals _withdrawalProof[]: 600,
```
Got some data on the bounds here if it helps: https://dune.com/queries/3433954/5768623

Based on that, the min proof length is 5 and the max we've seen is 8, so we should bump that length to 9 or 10 to be conservative there. The element lengths range from 32 to 532 bytes, so using 32 and 600 as the bounds there would be great

JuanCoRo replied:
Given that, for the time being, we cannot enforce bounds on the elements of `bytes[]`, what would be your preferred approach? Set the concrete length to 600?

mds1 replied:
Yea I think setting to element length to 600 bytes is good

mds1 committed:
ci: run kontrol on develop and allow manual dispatch

mds1 committed:
ci: rename parameter

JuanCoRo committed:
`OptimismPortalKontrol`: add remaining ranges for `_withdrawalProof`

JuanCoRo committed:
Add forge-like UX to `run-kontrol.sh`

JuanCoRo committed:
`pausability-lemmas.k`: clean file

Commit:
general tests, further lemmas, summary claim

JuanCoRo committed:
Address shellcheck failures

JuanCoRo committed:
Change `pausability-lemmas.k` to `pausability-lemmas.md`

mds1 commented on packages/contracts-bedrock/test/kontrol/pausability-lemmas.md:
````diff
@@ -0,0 +1,234 @@
+```
+
+Lemmas
+------
````
This renders as a header section with no content, should this be removed?

JuanCoRo replied:
Initially added it to have some general comments about lemmas, but the pointed-out documentation should cover that. Removed it is!

mds1 commented on packages/contracts-bedrock/test/kontrol/pausability-lemmas.md:
```diff
@@ -0,0 +1,234 @@
+
+## Arithmetic
+
+Lemmas on arithmetic reasoning.
```
Some of these seem universally true, so just curious why the need to be specified here instead of being natively part of kontrol?

JuanCoRo replied:
These general enough lemmas could be part of future versions of Kontrol. Most likely, these lemmas are separate from Kontrol because, in the past, we did not encounter the need to add them.
Needing them in this project is a solid reason to add them to Kontrol, although it is not clear if/when this will be done, hence why we keep them in this project's specific lemmas

mds1 replied:
Ah interesting, thanks 👌

mds1 commented on packages/contracts-bedrock/test/kontrol/pausability-lemmas.md:
```diff
@@ -0,0 +1,234 @@
+    //
+
+    // Invertibility of #buf and #asWord
+    // TODO: remove once the KEVM PR is merged
```
Can we link to the PR in this comment?

JuanCoRo replied:
These lemmas are not necessary anymore since they've been included in Kontrol. I've removed them

mds1 commented on packages/contracts-bedrock/test/kontrol/pausability-lemmas.md:
```diff
@@ -0,0 +1,234 @@
+        [concrete(S, W, Y), simplification]
+
+    // #asWord is equality
+    // TODO: remove once the KEVM PR is merged
```
Same, can we link to the PR in this comment too?

JuanCoRo replied:
Same as with the other lemma, just removed it

mds1 commented on packages/contracts-bedrock/test/kontrol/pausability-lemmas.md:
```diff
@@ -0,0 +1,234 @@
+
+## `copy_memory_to_memory` summary
+
+The following rule is a summarization of the `copy_memory_to_memory` function. This function is automatically generated by the compiler. In it's YUL form, is as follows:
```
```suggestion
The following rule is a summarization of the `copy_memory_to_memory` function. This function is automatically generated by the Solidity compiler. In it's Yul form, it is as follows:
```

mds1 commented on packages/contracts-bedrock/test/kontrol/pausability-lemmas.md:
```diff
@@ -0,0 +1,234 @@
+
+It is used to copy `length` bytes of memory from index `src` to index `dest`, doing so in steps of 32 bytes, and right-padding with zeros to a multiple of 32.
+
+We need to enforce some limit on the length of bytearrays and indices into bytearrays in order to avoid chop-reasoning.
```
Can we add a brief summary of chop reasoning here and why we want to avoid it? Ref https://github.com/ethereum-optimism/optimism/pull/9530#discussion_r1491340752

JuanCoRo replied:
Avoiding chop reasoning is more of a consequence of enforcing a limit on the length of the byte arrays, which is really introduced to satisfy compiler-introduced checks on the lengths of byte arrays. It has been documented as such in [this section](https://github.com/runtimeverification/_audits_Ethereum-optimism_pausability/blob/kontrol-symbolic-bytes/packages/contracts-bedrock/test/kontrol/pausability-lemmas.md#copy_memory_to_memory-summary). The whole file has been updated with what aims to be the final documentation. However, I could still add some words about chop reasoning if you'd like!

mds1 commented on packages/contracts-bedrock/test/kontrol/pausability-lemmas.md:
```diff
@@ -0,0 +1,234 @@
+Summaries
+---------
+
+Summary functions are rewrite rules that encapsulate the effects of executing a function. Thus, instead of executing the function itslef, Kontrol will just apply the summary rule.
```
```suggestion
Summary functions are rewrite rules that encapsulate the effects of executing a function. Thus, instead of executing the function itself, Kontrol will just apply the summary rule.
```

mds1 commented on packages/contracts-bedrock/test/kontrol/pausability-lemmas.md:
```diff
@@ -0,0 +1,234 @@
+
+## `#asWord`
+
+Lemmas about [`#asWord`](https://github.com/runtimeverification/evm-semantics/blob/master/kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md#bytes-helper-functions). `#asWord` will interperet a stack of bytes as a single word (with MSB first).
```
```suggestion
Lemmas about [`#asWord`](https://github.com/runtimeverification/evm-semantics/blob/master/kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md#bytes-helper-functions). `#asWord` will interpret a stack of bytes as a single word (with MSB first).
```

mds1 commented on packages/contracts-bedrock/test/kontrol/pausability-lemmas.md:
```diff
@@ -0,0 +1,234 @@
+
+## `#asInteger`
+
+Lemmas about [`#asInteger`](https://github.com/runtimeverification/evm-semantics/blob/master/kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md#bytes-helper-functions). `#asInteger` will interperet a stack of bytes as a single arbitrary-precision integer (with MSB first).
```
```suggestion
Lemmas about [`#asInteger`](https://github.com/runtimeverification/evm-semantics/blob/master/kevm-pyk/src/kevm_pyk/kproj/evm-semantics/evm-types.md#bytes-helper-functions). `#asInteger` will interpret a stack of bytes as a single arbitrary-precision integer (with MSB first).
```

mds1 commented on packages/contracts-bedrock/test/kontrol/pausability-lemmas.md:
Pretty neat you can define the lemmas this way in a `.md` file instead of a `.k` file

mds1 commented on .circleci/config.yml:
```diff
@@ -7,6 +7,16 @@ parameters:
+  #       - equal: [ "develop", <<pipeline.git.branch>> ]
+  #       - equal: [ true, <<pipeline.parameters.kontrol_dispatch>> ]
+  # Add a new `*_dispatch` parameter for any pipeline you want manual dispatch for.
+  kontrol_dispatch:
```
We ended up using this same trick in other places, you'll need to merge in the latest develop to resolve the resulting conflict in this file (ref https://github.com/ethereum-optimism/optimism/pull/9654)

JuanCoRo replied:
See [this commit](https://github.com/ethereum-optimism/optimism/pull/9530/commits/d580048caabf88a08c9bba39f943ee7f01873bb7) for the conflict resolution. Let me know if that looks alright to you!

mds1 commented on packages/contracts-bedrock/test/kontrol/scripts/common.sh:
```diff
@@ -3,12 +3,18 @@
   echo "  local              Run locally, enforces registered versions.json version for better reproducibility. (Recommended)" 1>&2
   echo "  dev                Run locally, does NOT enforce registered version. (Useful for developing with new versions and features)" 1>&2
+  echo "" 1>&2
+  echo "Tests executed:"
+  echo "  script             Execute the tests recorded in run-kontrol.sh" 1>&2
+  echo "  TESTS              Execute the tests provided as arguments" 1>&2
```
We probably should pull the `usage` function out into the respective shell scripts, since these arguments are presumably not valid for `make-deployment-summary.sh`, but since this is in `common` the helps shows them as valid

JuanCoRo replied:
Perhaps the `usage` function should remain in `common.sh`, since `usage` can be accessed from `make-deployment-summary.sh` and `run-kontrol.sh` via `parse_first_arg` (some refactoring needed atm in  `make-deployment-summary.sh`, since right now doesn't accept the `help` argument). We can have specialized `usage` functions for each script and a general one that calls the correct `usage_*` by reading a flag of where the `usage` call comes from. Do you have something else in mind?

mds1 replied:
That works for me!

mds1 commented on packages/contracts-bedrock/test/kontrol/scripts/run-kontrol.sh:
```diff
@@ -105,13 +106,29 @@ regen=
+  test_list=( # "OptimismPortalKontrol.prove_proveWithdrawalTransaction_paused0" \
+              # "OptimismPortalKontrol.prove_proveWithdrawalTransaction_paused1" \
+              "OptimismPortalKontrol.prove_proveWithdrawalTransaction_paused2" \
+              # "OptimismPortalKontrol.prove_proveWithdrawalTransaction_paused3" \
```
Shellcheck error here: https://www.shellcheck.net/wiki/SC1143

Impliciation is that if I uncomment `# "OptimismPortalKontrol.prove_proveWithdrawalTransaction_paused4" \` a few lines down it won't actually run since comments in this form terminate the command. That linked shellcheck page explains the workaround

JuanCoRo replied:
I've removed the unexecuted lemmas out of the `test_list` so that the shellcheck error si not present anymore. Will add them back before merging this PR tho!

mds1 commented on packages/contracts-bedrock/test/kontrol/pausability-lemmas.md:
```diff
@@ -0,0 +1,234 @@
+Kontrol Lemmas
```
_(arbitrary comment location to start a thread)_

The README still references `pausability-lemmas.k` twice, we'll have to update that (h/t @coderabbitai in https://github.com/ethereum-optimism/optimism/pull/9530#discussion_r1508016880)

JuanCoRo replied:
Updated it accordingly :ok_hand:

JuanCoRo committed:
`versions.json`: bump `kontrol` from `0.1.156` to `0.1.178`

JuanCoRo committed:
`OptimismPortalKontrol`: update `kontrol` natspec to version 0.1.178

JuanCoRo committed:
`pausability-lemmas.md`: remove unused `Lemmas` header

JuanCoRo committed:
Update packages/contracts-bedrock/test/kontrol/pausability-lemmas.md

Co-authored-by: Matt Solomon <matt@mattsolomon.dev>

JuanCoRo committed:
Update packages/contracts-bedrock/test/kontrol/pausability-lemmas.md

Co-authored-by: Matt Solomon <matt@mattsolomon.dev>

JuanCoRo committed:
Update packages/contracts-bedrock/test/kontrol/pausability-lemmas.md

Co-authored-by: Matt Solomon <matt@mattsolomon.dev>

JuanCoRo committed:
Update packages/contracts-bedrock/test/kontrol/pausability-lemmas.md

Co-authored-by: Matt Solomon <matt@mattsolomon.dev>

JuanCoRo committed:
`pausability-lemmas.md`: fix spelling typo

JuanCoRo committed:
`pausability-lemmas.md`: fix typo `bytearrays` to `byte arrays`

JuanCoRo committed:
`run-kontrol.sh`: correctly format temorarily unexecuted lemmas

JuanCoRo committed:
`common.sh`: execute `copy_to_docker` only when in docker mode

JuanCoRo committed:
`make-summary-deployment.sh`: add argument check before parsing

JuanCoRo committed:
Reflect `kontrol summary` change to `kontrol load-state-diff`

From version `0.1.162`, `kontrol summary` has been renamed to `kontrol
load-state-diff`.
The reason of this renaming is that `kontrol summary` will be
used by kontrol's compositional symbolic execution.
Also, changing the name to `load-state-diff` makes more explicit what the
command does.
Related PR: https://github.com/runtimeverification/kontrol/pull/374

JuanCoRo committed:
`pausability-lemmas.md`: remove upstreamed lemmas

JuanCoRo committed:
fix: writing typos

Commit:
lemma text pass

JuanCoRo commented on packages/contracts-bedrock/test/kontrol/pausability-lemmas.md:
```diff
@@ -0,0 +1,219 @@
+Kontrol Lemmas
+==============
+
+Lemmas are K rewrite rules that enhance the reasoning power of Kontrol. For more information on lemmas, please consult [this section](https://docs.runtimeverification.com/kontrol/guides/advancing-proofs) of the Kontrol documentation.
+
+This file contains the lemmas required to run the proofs included in the [proofs](./proofs) folder. Some of these lemmas are general enough to likely be incorporated into future versions of Kontrol, while others are specific to the challenges presented by the proofs.
+
+Similarly to other files such as [`cheatcodes.md`](https://github.com/runtimeverification/kontrol/blob/master/src/kontrol/kdist/cheatcodes.md), we use the idiomatic way of programming in Kontrol, which is [literate programming](https://en.wikipedia.org/wiki/Literate_programming), allowing for better documentation of the code.
```
Kontrol is described in the readme of the project's root folder, so there's no need of describing it again in the lemmas file

Commit:
paragraph about summary maintainability

JuanCoRo committed:
README.md: include latest changes

JuanCoRo committed:
pausability-lemmas.md: markdown link typo

JuanCoRo committed:
KontrolUtils: add documentation comment

mds1 commented on versions.json:
```diff
@@ -4,5 +4,5 @@
   "nvm": "v20.9.0",
   "slither": "0.10.0",
-  "kontrol": "0.1.127"
+  "kontrol": "0.1.178"
```
Does this version include the JUint XML reports from https://github.com/runtimeverification/kontrol/issues/329? Would be great to include that in this PR if we can. Relevant CircleCI docs [here](https://circleci.com/docs/collect-test-data/), and we currently use that for our Go tests if you want to see the implementation in the existing CI file

JuanCoRo replied:
That Kontrol version doesn't include it, but we can add it without a problem! My only question is, in light of offloading computation, would it still make sense to add the reports to this PR? Or would it be best to wait? Not sure how the CI landscape will change

mds1 replied:
My current understanding we should open the "offload compute" PR into this PR before merging, because if we merge this PR our CI will break due to resource exhaustion. But we can still use those reports per [this comment](https://oplabs-pbc.slack.com/archives/C04BP7VL6DC/p1709747090575859?thread_ts=1709662100.625049&cid=C04BP7VL6DC)

JuanCoRo replied:
We can modify which tests your machines will execute from this PR so that resources will not be exhausted. This would allow us to merge this PR first, having the final code in place, and not have to wait to have the "offload compute" PR merged first. We can do it either way, just wanted to note that we can set it to execute all tests later

mds1 replied:
Ok perfect, I'm good with that approach. So the only remaining blockers for merge here would be changing the set of tests to run, and adding the JUnit XML reports for those?

JuanCoRo replied:
I believe so! I didn't get to do the JUnit XML reports today. Will try to get them done tomorrow!

JuanCoRo replied:
An update on this, we're pending on one PR to support generating the JUint XML reports with tests in nested dirs. Tomorrow should be ready, and after that, everything will be ready to go!

JuanCoRo replied:
The last commits should contain everything related to this conversation

JuanCoRo committed:
pausability-lemmas.md: fix markdown typo vol2

JuanCoRo committed:
pausability-lemmas.md: fix markdown typo vol3

JuanCoRo committed:
Add specialized usage functions

JuanCoRo committed:
Merge remote-tracking branch 'origin/develop' into kontrol-symbolic-bytes

JuanCoRo committed:
`README.md`: remove `bash` in `make-summary-deployment.sh` description

JuanCoRo committed:
`README.md`: complete `Add New Proofs` section

JuanCoRo committed:
versions.json: bump kontrol from 0.1.178 to 0.1.196

JuanCoRo committed:
run-kontrol.sh: add `--xml-test-report` flag

JuanCoRo committed:
.gitignore: add `kontrol_prove_report.xml`

JuanCoRo committed:
foundry.toml: set `ast = true` in `kprove` profile

JuanCoRo committed:
config.yml: set correct path for kontrol `store_artifacts`

JuanCoRo committed:
config.yml: add `store_test_results` step to `kontrol-tests` job

JuanCoRo committed:
package.json: execute `run-kontrol.sh` with `script` option

We run `run-kontrol.sh` with the `script` option to avoid executing all proofs

JuanCoRo committed:
run-kontrol.sh: remove proof with 10 elements in favor of 0 and 1

The longer the `_withdrawalProof` array the longer the execution time
Adding the lengths 0 and 1 fits within the max cpus and won't take as long to run

JuanCoRo committed:
Merge remote-tracking branch 'origin/develop' into kontrol-symbolic-bytes

mds1 committed:
chore: typos and formatting

ajsutton commented on .circleci/config.yml:
```diff
@@ -2063,9 +2068,11 @@ workflows:
+  develop-kontrol-tests:
     when:
-      equal: [ build_four_hours, <<pipeline.schedule.name>> ]
+      or:
+        - equal: [ "develop", <<pipeline.git.branch>> ]
+        - equal: [ true, <<pipeline.parameters.kontrol_dispatch>> ]
```
This will wind up running for every scheduled trigger as well.  Not sure I'll get the indenting right here, but it should be like the `develop-fault-proofs` job above.
```suggestion
      and:
        - or:
          - equal: [ "develop", <<pipeline.git.branch>> ]
          - equal: [ true, <<pipeline.parameters.kontrol_dispatch>> ]
        - not:
            equal: [ scheduled_pipeline, << pipeline.trigger_source >> ]
```

mds1 replied:
Ah thank you, fixed in d702f2072 (ref https://github.com/ethereum-optimism/optimism/pull/9654)

ajsutton reviewed (approved):
Approved, but need to add the filter to ensure Kontrol doesn't run on every scheduled trigger.

mds1 committed:
ci: fix kontrol trigger

JuanCoRo committed:
README.md: minor typo
